### ISSUE TITLE
ParallelSSHClient: run_command on selective host from available hosts

### ISSUE THREAD

**Issue by tripples** [pkittenis/parallel-ssh#51]
Hi,

This is bit modification in run_command so that we can execute command on selective hosts rather than all(required in my use cases). Please let me know your opinion about this.

I modified run_command bit to add host filter hook to try it.

```python
    def run_command(self, *args, **kwargs):
        stop_on_errors = kwargs.pop('stop_on_errors', True)
        host_filter = kwargs.pop('host_filter', None)
        if host_filter:
            filtered_hosts = host_filter(tuple(self.hosts))
        else:
            filtered_hosts = self.hosts
        cmds = [self.pool.spawn(self._exec_command, host, *args, **kwargs)
                for host in filtered_hosts]
        output = {}
        for cmd in cmds:
            try:
                self.get_output(cmd, output)
            except Exception, ex:
                if stop_on_errors:
                    raise ex
        return output
```
Host filter can be some function to filter hosts. Some calling conventions can be set to this hook
so that it does not alter ParallelSSHClient things.
```python
def first_node(hosts):
	return [hosts[0], ]
```
Can be executed like,

```python
client.run_command("ls", host_filter=first_node)
```

Regards,
Sanket

**Comment by pkittenis** [comment]
Wouldn't 

```
hosts = [..]
client = ParallelSSHClient(filter(my_filter_func, hosts))
client.run_command(..)
``` 

suffice for the use case? Functionally there is no difference to the above changes - the client will still maintain existing connections to hosts previously connected to, as opposed to creating a new client for each bunch of hosts for example.

Not opposed to using filters as such but applying them on `run_command` makes parallel client behaviour non-deterministic and the client's output will not match the hosts commands actually ran on. 

That use case would perhaps be better served by manipulating the host list as given to the client rather than applying a filter on top of it on each execution of ```run_command```.

Anything that is an iterator can be used for host list mind, not just lists, so generator expressions et al are all valid.

With a basic substring search filter function, the code could be written like this:

```
my_filter_func = lambda host: host if host.find('substring') else None

hosts = [..]
client = ParallelSSHClient(filter(my_filter_func,hosts))
client.run_command(...)
```

To change the filter either change `client.hosts` to point to new ```filter(new_filter_func,hosts)``` or change the filter function directly.

Pros:
 * Deterministic behaviour - every call to ```run_command``` executes the same thing on the same hosts.
 * Output from client is only for host list commands actually ran on
 * Allows for dynamically changing host list using standard python language mechanics, iterators, generators etc.

Cons:
  * Arguably a bit more code to write when making the ```ParallelSSHClient``` object.

Make sense?

**Comment by tripples** [comment]
Agree with your explanation. I guess your approach makes more sense .  Though it is very intuitive, it will be good if we can write short note in doc/FAQ.

I also had few queries, not related to this thread. When  run_command is called it blocks until context does not not return from called function,

```   
run_command doc
This function will block until all commands have **started** and
        then return immediately. Any connection and/or authentication exceptions
        will be raised here and need catching.
```
I am just confused with meaning of **started** here.

I have commands c1, c2, c3 which does not depend on each other.  How do make sure I add them to pool and run simultaneously ?  Do I need to create different parallelsshclient and my application maintains required parallelism?

I tried each of following command, each command takes time to return depending on sleep time.
```
In [17]: c1 = p.run_command("sleep 15") # takes around 15 sec

In [18]: c2 = p.run_command("sleep 15")  

In [19]: c3 = p.run_command("sleep 60") # takes around 60 sec
```
Are they supposed to run/complete only when I call ```pool.join```. 

Of course there can be argument of deterministic or non-deterministic behavior based on dependencies between commands.

This returns immediately,
```
In [20]: c4 = p.run_command(''while true; do echo "Hello"; done')

In [61]: [line for line in c4['myhost']['stdout']]
# Above statement never returns.
```

Just to return commands immediately I called, which will return immediately, 
```
In [21]: o1 = p.run_command("sleep 15 &")

In [22]: o2 = p.run_command("sleep 20 &")

In [23]: o3 = p.run_command("sleep 60 &")

```
Also can we add timeout in run_command ?

Please let me know I am missing anything.

I just bumped into this project, some of things in task api are kind of things we discussed here.
http://clustershell.readthedocs.org/en/latest/api/Task.html
https://github.com/cea-hpc/clustershell

I guess fundamental difference is
```
ParallelSSH's design goals and motivation are to provide a library for running asynchronous SSH 
commands in parallel with no load induced on the system by doing so with the intended usage being 
completely programmatic and non-interactive
```

**Comment by pkittenis** [comment]
| I am just confused with meaning of started here.

Currently means either channel is ready to receive data or channel is closed/command finished. Your example shows an edge case where if a command has no output the channel never goes into ready to receive state making `run_command` block for the command's duration, like in your sleep examples. Raised #56 for it.

Have made a few changes to resolve this by starting cmds asynchronously and also make starting commands faster. It means slightly backwards incompatible code so there will likely be a new major version to include them. 

Breaking change from async cmd start is previously exit code for commands that return immediately would be available in output without having to call either `client.join`, `client.get_exit_codes` or iterate over stdout/stderr (which implies cmd completion). Now at least one of these is required. This is actually covered in documentation but code that previously relied on this behaviour will break.

With changes, due in master branch shortly.

```
from pssh import ParallelSSHClient
client = ParallelSSHClient(['localhost'])
from datetime import datetime
datetime.now()

datetime.datetime(2016, 2, 23, 15, 42, 15, 974437)

client.run_command('sleep 10')

datetime.now()

datetime.datetime(2016, 2, 23, 15, 42, 17, 798583)
```

|  How do make sure I add them to pool and run simultaneously ?

```
for cmd in ['uname', 'whoami']:
    client.run_command(cmd)
```

See `pool_size` [at docs](http://parallel-ssh.readthedocs.org/en/latest/pssh_client.html#pssh.pssh_client.ParallelSSHClient) for level of parallelism. 

The first cmd will actually take slightly longer to start due to having to establish a new connection. Subsequent cmds use existing connection and will start as fast at the connection/remote server allow, assuming `pool_size` is >= number of hosts.

| c4 = p.run_command(''while true; do echo "Hello"; done')
| [line for line in c4['myhost']['stdout']]
| # Above statement never returns.

Not surprising :). Command output is infinite. Gathering a list of infinite output will never finish, or at least until the process gets killed for taking up too much memory.

Setting timeout is #35 and fairly easy to do if you'd be up for it :) 

For now can set `channel.timeout = <timeout>` where channel is from `output[<host>]['channel']`.

Regarding cluster shell, that seems to be a tool for administering clusters of servers, similar to capistrano and fabric.

It sounds like a perfect use case for parallel ssh, as in cluster shell can use parallel-ssh to run its commands asynchronously rather than spawning ssh processes (in a shell) all over the place. 

Parallel-SSH does not, however, have a goal to replicate that a task-based API or any kind of cmd line interface - that is out of scope for this project. You are correct in the fundamental differences. It's an asynchronous parallel ssh client library, simple as that. Other tools can use the library to do those kinds of things.

Try running one cmd on 1000 servers at the same time with cluster shell or any other non-async tool, see what happens :)
